File: DotNetPoly/Fields/HouseField.vb
Namespace Fields

    Public Class HouseField
        Inherits BaseField

        Friend Sub New(pXmlNode As Xml.XmlNode, pSettings As Settings)
            MyBase.New(pXmlNode.Attributes("name")?.InnerText)
            Integer.TryParse(pXmlNode.Attributes("group")?.InnerText, Group)

            Dim upgradeNodes = pXmlNode.SelectNodes("./level")
            ReDim UpgradeMatrix(upgradeNodes.Count - 1, 1)

            For i = 0 To upgradeNodes.Count - 1
                    Integer.TryParse(upgradeNodes(i).Attributes("cost").InnerText, UpgradeMatrix(i, 0))
                    Integer.TryParse(upgradeNodes(i).Attributes("rent").InnerText, UpgradeMatrix(i, 1))
                Next
            maxUpgradeLevel = upgradeNodes.Count - 1


            UpgradeAlgorithmRent = pSettings.RentUpgradeMethode
            UpgradeAlgorithmCost = pSettings.CostUpgradeMethode
            UpgradePreReq = pSettings.UpgradePreReq
        End Sub

        Private ReadOnly maxUpgradeLevel As Integer = 2

        Private ReadOnly UpgradePreReq As eUpgradeAlgorithm
        Private ReadOnly UpgradeAlgorithmRent As eUpgradeAlgorithm
        Private ReadOnly UpgradeAlgorithmCost As eUpgradeAlgorithm

        Public Enum eUpgradeAlgorithm
            FullGroup

            Table
            FullGroupDoublesRent
        End Enum

        Private ReadOnly UpgradeMatrix(,) As Integer

        ReadOnly Property Cost As Integer
            Get
                Dim myCost As Integer = 0
                For i = 0 To UpgradeLevel
                    myCost += UpgradeMatrix(i, 0)
                Next
                Return myCost
            End Get
        End Property

        ReadOnly Property UpgradeLevel As Integer = 0

        Public ReadOnly Group As Integer

        ReadOnly Property Rent As Integer
            Get
                Dim multiplier As Integer
                If UpgradeAlgorithmRent = eUpgradeAlgorithm.FullGroupDoublesRent AndAlso Not Me.Owner.Equals(GameBoard.BANK) AndAlso hasCompleteStreet(Me.Owner) Then
                    multiplier = 2
                Else
                    multiplier = 1
                End If

                Return UpgradeMatrix(UpgradeLevel, 1) * multiplier
            End Get
        End Property

        Friend Sub Buy(pPlayer As BasePlayer)
            Me.GameBoard.RaiseChangeOwner(Me, pPlayer)

            pPlayer.TransferMoney(Me.Owner, Me.Cost)
            ChangeOwner(pPlayer)

        End Sub

        Private Function hasCompleteStreet(pPlayer As Entity) As Boolean
            If GameBoard.Statistics.CompleteGroups = 0 Then Return False
            For Each h As HouseField In GameBoard.Groups(Me.Group)
                If Not h.Owner.Equals(pPlayer) Then
                    Return False
                End If
            Next
            Return True
        End Function

        Public Function CanUpgrade(pPlayer As BasePlayer) As Boolean
            If UpgradeLevel >= maxUpgradeLevel Then Return False
            If UpgradeAlgorithmRent = eUpgradeAlgorithm.Table AndAlso UpgradeLevel >= UpgradeMatrix.GetLength(0) Then Return False
            If pPlayer.Cash < UpgradeCost Then Return False

            'If pPlayer.GetType() Is GetType(Entity) Then Return False
            If UpgradePreReq = eUpgradeAlgorithm.FullGroup Then
                Return hasCompleteStreet(pPlayer)
            End If
            Return True
        End Function

        Public ReadOnly Property UpgradeCost As Integer
            Get
                Select Case UpgradeAlgorithmRent
                    Case eUpgradeAlgorithm.Table
                        Return UpgradeMatrix(UpgradeLevel, 0)
                    Case Else 'eUpgradeAlgorithm.Linear
                        Return Math.Max(CInt(Me.Cost * (_UpgradeLevel + 1) / 2), _UpgradeLevel + 1) 'Cost >= 1
                End Select
            End Get
        End Property

        Friend Sub Upgrade()
            Dim _owner = DirectCast(Owner, BasePlayer)
            Owner.TransferMoney(GameBoard.BANK, UpgradeCost)
            If _owner.IsPlayering Then
                DirectCast(Owner, BasePlayer).RaiseEventHouseUpgrade(Me)
                _UpgradeLevel += 1
            End If

            If Me.Cost > GameBoard.Statistics.MaxCost Then GameBoard.Statistics.MaxCost = Me.Cost
            If Me.Rent > GameBoard.Statistics.MaxRent Then GameBoard.Statistics.MaxRent = Me.Rent
            If Me.UpgradeLevel > GameBoard.Statistics.MaxUpgradeLevel Then GameBoard.Statistics.MaxUpgradeLevel = Me.UpgradeLevel
        End Sub

        Friend Overrides Sub onMoveOn(pPlayer As BasePlayer)
            MyBase.onMoveOn(pPlayer)

            If Me.Owner.Equals(GameBoard.BANK) Then
                pPlayer.onDelegateControl(New eActionType() {eActionType.BuyHouse, eActionType.Pass}, New BaseField() {Me}, Nothing, 0)
            ElseIf Me.Owner.Equals(pPlayer) Then
                If CanUpgrade(pPlayer) Then
                    pPlayer.onDelegateControl(New eActionType() {eActionType.UpgradeHouse, eActionType.Pass}, New BaseField() {Me}, Nothing, 0)
                End If

            Else
                pPlayer.onDelegateControl(eActionType.PayRent, Me)

            End If
        End Sub

        Friend Sub ChangeOwner(pPlayer As Entity)
            Dim dbg1 As Integer = 0

            If Me._owner.Equals(GameBoard.BANK) Then
                GameBoard.Statistics.SoldHouseFields += 1
            Else
                GameBoard.Statistics.SoldHouseFields -= 1
            End If





            Me._owner._ownfields.Remove(Me)
            Me._owner = pPlayer

            'If pPlayer.GetType() IsNot GetType(Entity) Then
            pPlayer._ownfields.Add(Me)
            'End If

            'HACK: Correcter for Wrong Count
            Dim totalsold As Integer
            For Each f As HouseField In GameBoard._Housefields.Values
                If Not f.Owner.Equals(GameBoard.BANK) Then
                    totalsold += 1
                End If
            Next

            If Me.Cost > GameBoard.Statistics.MaxCost Then GameBoard.Statistics.MaxCost = Me.Cost
            If Me.Rent > GameBoard.Statistics.MaxRent Then GameBoard.Statistics.MaxRent = Me.Rent
            If Me.UpgradeLevel > GameBoard.Statistics.MaxUpgradeLevel Then GameBoard.Statistics.MaxUpgradeLevel = Me.UpgradeLevel

            'Check Whether Group ist complete
            Dim groupcomplete As Boolean = True
            For Each h As HouseField In GameBoard.Groups(Me.Group)
                If Not h.Owner.Equals(pPlayer) Then
                    groupcomplete = False
                End If
            Next

            If groupcomplete Then
                GameBoard.Statistics.CompleteGroups = 0
                'If Group is complete, calculate all Groups
                For Each g In GameBoard.Groups
                    Dim prevOwner As Entity = Nothing
                    groupcomplete = False
                    For Each h As HouseField In g
                        If (prevOwner IsNot Nothing AndAlso Not h.Owner.Equals(prevOwner)) OrElse h.Owner.Equals(GameBoard.BANK) Then
                            groupcomplete = False
                            Exit For
                        Else
                            groupcomplete = True
                            prevOwner = h.Owner
                        End If
                    Next
                    If groupcomplete Then
                        GameBoard.Statistics.CompleteGroups += 1
                    End If
                Next
                GameBoard.Statistics.CompleteGroups += 1
            End If


            Diagnostics.Debug.WriteLineIf(GameBoard.Statistics.SoldHouseFields <> totalsold, $"HACK: SoldHouseFields irregular in ChangeOwner {GameBoard.Statistics.SoldHouseFields} <> {totalsold}")
            GameBoard.Statistics.SoldHouseFields = totalsold

        End Sub

    End Class
End Namespace

File: DotNetPoly/Fields/HouseField.cs
using System.Linq;

namespace DotNetPoly
{
    namespace Fields
    {
        public class HouseField : Field
        {
            internal HouseField(System.Xml.XmlNode pXmlNode, Settings pSettings) : base(pXmlNode.Attributes["name"]?.InnerText)
            {
                ushort.TryParse(pXmlNode.Attributes["group"]?.InnerText, out _Street);

                var upgradeNodes = pXmlNode.SelectNodes("./level");
                UpgradeMatrix = new ushort[upgradeNodes.Count - 1 + 1, 2];

                RentUpgradeFactor = pSettings.DefaultRentUpgradeFactor;
                var loopTo = upgradeNodes.Count - 1;
                for (var i = 0; i <= loopTo; i++)
                {
                    ushort.TryParse(upgradeNodes[i].Attributes["cost"].InnerText, out UpgradeMatrix[i, 0]);
                    ushort.TryParse(upgradeNodes[i].Attributes["rent"].InnerText, out UpgradeMatrix[i, 1]);
                    try
                    {
                        int.TryParse(upgradeNodes[i].Attributes["rent"].Attributes["factor"].InnerText, out RentUpgradeFactor);
                    }
                    catch
                    {
                        RentUpgradeFactor = pSettings.DefaultRentUpgradeFactor;
                    }
                }
                maxUpgradeLevel = upgradeNodes.Count;

                UpgradeAlgorithmRent = pSettings.RentUpgradeMethode;
                UpgradeAlgorithmCost = pSettings.CostUpgradeMethode;
            }


            private readonly int maxUpgradeLevel = 2;

            private readonly int RentUpgradeFactor;

            private readonly eUpgradeAlgorithm UpgradeAlgorithmRent;
            private readonly eUpgradeAlgorithm UpgradeAlgorithmCost;

            public bool HasCredit { get; private set; }
            public enum eUpgradeAlgorithm
            {
                FullGroup,
                Table,
                OnGroup,
                OnHouse
            }

            // When could a field be upgraded
            public enum eUpgradeRange
            {
                OnThisHouse     // only when positioned on House
    ,
                OnThisGroup     // every house in positioned group
    ,
                OnOwnField      // all, when on own field
    ,
                Always          // always
            }

            private readonly ushort[,] UpgradeMatrix;

            public ushort Cost
            {
                get
                {
                    ushort myCost = 0;
                    for (var i = 0; i <= UpgradeLevel; i++)
                        myCost += UpgradeMatrix[i, 0];
                    return myCost;
                }
            }

            public int UpgradeLevel { get; private set; } = 0;

            // TODO: Whats about no Street
            internal readonly ushort _Street;

            public ushort Rent
            {
                get
                {
                    ushort myrent;

                    if (UpgradeAlgorithmRent == eUpgradeAlgorithm.FullGroup && !this.Owner.Equals(GameBoard.BANK) && HasCompleteStreet(this.Owner))
                        myrent = System.Convert.ToUInt16(UpgradeMatrix[UpgradeLevel, 1] * 2);
                    else
                        myrent = System.Convert.ToUInt16(UpgradeMatrix[UpgradeLevel, 1]);

                    if (!this.Owner.Equals(GameBoard.BANK))
                        myrent *= System.Convert.ToUInt16(((Player)Owner).RentPercentMultiplier / (double)100);

                    return myrent;
                }
            }

            internal void Buy(Player pPlayer)
            {
                Buy(pPlayer, Cost);
            }
            [System.Obsolete()]
            internal void Buy(Player pPlayer, uint pCost)
            {
                GameBoard.RaiseChangeOwner(this, pPlayer);

                pPlayer.TransferMoney(this.Owner, pCost);
                ChangeOwner(pPlayer);
            }

            private bool HasCompleteStreet(Entity pPlayer)
            {
                if (GameBoard.Statistics.CompleteGroups == 0)
                    return false;

                return (from h in GameBoard.Groups[this._Street] select h).All((HouseField h) => h.Owner.Equals(pPlayer));
                foreach (HouseField h in GameBoard.Groups[this._Street])
                {
                    if (!h.Owner.Equals(pPlayer))
                        return false;
                }
                return true;
            }


            public bool CanUpgrade(Player pPlayer)
            {
                // TODO: Check for street
                foreach (var h in new HouseField[] { this })
                {
                    if (h.UpgradeLevel >= h.maxUpgradeLevel)
                        return false;
                    if (h.UpgradeAlgorithmRent == eUpgradeAlgorithm.Table && h.UpgradeLevel >= h.UpgradeMatrix.GetLength(0))
                        return false;
                    if (pPlayer.Cash < h.UpgradeCost)
                        return false;
                }

                return true;
            }

            public ushort UpgradeCost
            {
                get
                {
                    switch (UpgradeAlgorithmRent)
                    {
                        case eUpgradeAlgorithm.Table:
                            {
                                return UpgradeMatrix[UpgradeLevel, 0];
                            }

                        default:
                            {
                                return System.Convert.ToUInt16(System.Math.Max((this.Cost * (UpgradeLevel + 1) / (double)2), UpgradeLevel + 1)); // Cost >= 1
                            }
                    }
                }
            }

            internal void Upgrade()
            {
                if (Owner is Player PlayerOwner)
                {
                    PlayerOwner.TransferMoney(GameBoard.BANK, UpgradeCost);
                    if (PlayerOwner.IsPlayering)
                    {
                        PlayerOwner.RaiseEventHouseUpgrade(this);
                        UpgradeLevel += 1;
                    }

                    if (this.Cost > GameBoard.Statistics.MaxCost)
                        GameBoard.Statistics.MaxCost = this.Cost;
                    if (this.Rent > GameBoard.Statistics.MaxRent)
                        GameBoard.Statistics.MaxRent = this.Rent;
                    if (this.UpgradeLevel > GameBoard.Statistics.MaxUpgradeLevel)
                        GameBoard.Statistics.MaxUpgradeLevel = this.UpgradeLevel;
                }
            }

            //Shouldn return anything?!
            public Entity PayRent(Player pFromPlayer)
            {
                if (HasCredit)
                {
                    //No Rent, Bank Rent?
                    pFromPlayer.TransferMoney(GameBoard.BANK, Rent);
                    return GameBoard.BANK;
                }
                else
                {
                    pFromPlayer.TransferMoney(Owner, Rent);
                    return Owner;
                }
            }

            internal void TakeForCredit()
            {
                HasCredit = true;
                GameBoard.BANK.TransferMoney(this.Owner, this.Cost);
                GameBoard.Statistics.CreditHouses += 1;
            }

            internal void RemoveCredit()
            {
                Owner.TransferMoney(GameBoard.BANK, this.Cost);
                HasCredit = false;
                GameBoard.Statistics.CreditHouses -= 1;
            }

            internal override void onMoveOn(Player pPlayer)
            {
                base.onMoveOn(pPlayer);

                if (Owner.Equals(GameBoard.BANK))
                    pPlayer.onDelegateControl(new DelegateEventArgs(new eActionType[] { eActionType.BuyHouse, eActionType.Pass }, null, new Field[] { this }, 0));
                else if (this.Owner.Equals(pPlayer))
                {
                    if (CanUpgrade(pPlayer))
                        pPlayer.onDelegateControl(new DelegateEventArgs(new eActionType[] { eActionType.UpgradeHouse, eActionType.Pass }, null, new Field[] { this }, 0));
                }
                else
                    pPlayer.onDelegateControl(new DelegateEventArgs(eActionType.PayRent, this));
            }

            internal void ChangeOwner(Entity pPlayer)
            {
                if (Owner.Equals(GameBoard.BANK))
                    GameBoard.Statistics.SoldHouseFields += 1;
                else
                    GameBoard.Statistics.SoldHouseFields -= 1;

                Owner._ownfields.Remove(this.Name);
                Owner = pPlayer;

                pPlayer._ownfields.Add(this.Name,this);

                // HACK: Correcter for Wrong Count
                int totalsold = (from h in GameBoard._Housefields where !h.Value.Owner.Equals(GameBoard.BANK) select h.Value).Count();

                if (Cost > GameBoard.Statistics.MaxCost)
                    GameBoard.Statistics.MaxCost = this.Cost;
                if (Rent > GameBoard.Statistics.MaxRent)
                    GameBoard.Statistics.MaxRent = this.Rent;
                if (UpgradeLevel > GameBoard.Statistics.MaxUpgradeLevel)
                    GameBoard.Statistics.MaxUpgradeLevel = this.UpgradeLevel;

                // Check Whether Group ist complete
                

                if (HasCompleteStreet(pPlayer))
                {
                    GameBoard.Statistics.CompleteGroups = 0;
                    // If Group is complete, calculate all Groups
                    foreach (var g in GameBoard.Groups)
                    {
                        Entity prevOwner = null;
                        bool groupcomplete = false;
                        foreach (HouseField h in g)
                        {
                            if ((prevOwner != null && !h.Owner.Equals(prevOwner)) || h.Owner.Equals(GameBoard.BANK))
                            {
                                groupcomplete = false;
                                break;
                            }
                            else
                            {
                                groupcomplete = true;
                                prevOwner = h.Owner;
                            }
                        }
                        if (groupcomplete)
                            GameBoard.Statistics.CompleteGroups += 1;
                    }
                    GameBoard.Statistics.CompleteGroups += 1;
                }


                System.Diagnostics.Debug.WriteLineIf(GameBoard.Statistics.SoldHouseFields != totalsold, $"HACK: SoldHouseFields irregular in ChangeOwner {GameBoard.Statistics.SoldHouseFields} <> {totalsold}");
                GameBoard.Statistics.SoldHouseFields = totalsold;
            }
        }
    }
}
